src/Poseidon/SequencingSource.hs

```diff
@@ -0,0 +1,315 @@
+    | x == "half"  = pure SSFHalf
+    | x == "plus"  = pure SSFPlus
+    | otherwise    = fail $ "UDG " ++ show x ++ " not in [minus, half, plus]"
+
+instance Csv.ToField SSFUDG where
+    toField x = Csv.toField $ show x
+instance Csv.FromField SSFUDG where
+    parseField x = Csv.parseField x >>= makeSSFUDG
+instance ToJSON SSFUDG where
+    toEncoding x = text $ T.pack $ show x
+instance FromJSON SSFUDG where
+    parseJSON = withText "SSFUDG" (makeSSFUDG . T.unpack)
+
+-- |A datatype to represent Library_Built in a janno file
+data SSFLibraryBuilt =
+      SSFDS
+    | SSFSS
+    deriving (Eq, Ord, Generic, Enum, Bounded)
+
+instance Show SSFLibraryBuilt where
+    show SSFDS = "ds"
+    show SSFSS = "ss"
+
+makeSSFLibraryBuilt :: MonadFail m => String -> m SSFLibraryBuilt
+makeSSFLibraryBuilt x
+    | x == "ds"    = pure SSFDS
+    | x == "ss"    = pure SSFSS
+    | otherwise    = fail $ "Library_Built " ++ show x ++ " not in [ds, ss]"
+
+instance Csv.ToField SSFLibraryBuilt where
+    toField x = Csv.toField $ show x
+instance Csv.FromField SSFLibraryBuilt where
+    parseField x = Csv.parseField x >>= makeSSFLibraryBuilt
+instance ToJSON SSFLibraryBuilt where
+    toEncoding x = text $ T.pack $ show x
+instance FromJSON SSFLibraryBuilt where
+    parseJSON = withText "SSFLibraryBuilt" (makeSSFLibraryBuilt . T.unpack)
+
+-- | A data type to represent a seqSourceFile
+newtype SeqSourceRows = SeqSourceRows {getSeqSourceRowList :: [SeqSourceRow]}
+    deriving (Show, Eq, Generic)
+
+instance Semigroup SeqSourceRows where
+    (SeqSourceRows j1) <> (SeqSourceRows j2) = SeqSourceRows $ j1 `combineTwoSeqSources` j2
+        where
+        combineTwoSeqSources :: [SeqSourceRow] -> [SeqSourceRow] -> [SeqSourceRow]
+        combineTwoSeqSources seqSource1 seqSource2 =
+            let simpleSeqSourceSum = seqSource1 ++ seqSource2
+                toAddColNames = HM.keys (HM.unions (map (getCsvNR . sAdditionalColumns) simpleSeqSourceSum))
+                toAddEmptyCols = HM.fromList (map (\k -> (k, "n/a")) toAddColNames)
+            in map (addEmptyAddColsToSeqSourceRow toAddEmptyCols) simpleSeqSourceSum
+        addEmptyAddColsToSeqSourceRow :: Csv.NamedRecord -> SeqSourceRow -> SeqSourceRow
+        addEmptyAddColsToSeqSourceRow toAdd x =
+            x { sAdditionalColumns = CsvNamedRecord $ fillAddCols toAdd (getCsvNR $ sAdditionalColumns x) }
+        fillAddCols :: Csv.NamedRecord -> Csv.NamedRecord -> Csv.NamedRecord
+        fillAddCols toAdd cur = HM.union cur (toAdd `HM.difference` cur)
+
+instance Monoid SeqSourceRows where
+    mempty = SeqSourceRows []
+    mconcat = foldl' mappend mempty
+
+instance ToJSON SeqSourceRows where
+    toEncoding = genericToEncoding defaultOptions
+instance FromJSON SeqSourceRows
+
+-- | A data type to represent a row in the seqSourceFile
+-- See https://github.com/poseidon-framework/poseidon2-schema/blob/master/seqSourceFile_columns.tsv
+-- for more details
+data SeqSourceRow = SeqSourceRow
```

Comment:
Uff, would be super nice to somehow be able to reduce the code duplication with Janno, but I don't have much of an idea how. One could possibly define a type class `PoseidonTSV` or so, but I can see that being more complicated than useful.

Reply:
True. But keeping it separate gives us more flexibility and the large definitions between line 119 and 241 would be necessary anyway. Maybe we come back to it, when/if we consider to add another .tsv file to the standard (for whatever reason).